optimizer/evaluator/evaluator.go

```diff
@@ -849,9 +851,101 @@ func trimRight(str, remstr string) string {
 }
 
 func (e *Evaluator) funcDateArith(v *ast.FuncDateArithExpr) bool {
+	// health check for date and interval
+	nodeDate := v.Date.GetValue()
+	if types.IsNil(nodeDate) {
+		v.SetValue(nil)
+		return true
+	}
+	nodeInterval := v.Interval.GetValue()
+	if types.IsNil(nodeInterval) {
+		v.SetValue(nil)
+		return true
+	}
+	// parse date
+	fieldType := mysql.TypeDate
+	var resultField *types.FieldType
+	switch x := nodeDate.(type) {
+	case mysql.Time:
+		if (x.Type == mysql.TypeDatetime) || (x.Type == mysql.TypeTimestamp) {
+			fieldType = mysql.TypeDatetime
+		}
+	case string:
+		if !mysql.IsDateFormat(x) {
+			fieldType = mysql.TypeDatetime
+		}
+	case int64:
+		if t, err := mysql.ParseTimeFromInt64(x); err == nil {
+			if (t.Type == mysql.TypeDatetime) || (t.Type == mysql.TypeTimestamp) {
+				fieldType = mysql.TypeDatetime
+			}
+		}
+	}
+	if mysql.IsClockUnit(v.Unit) {
+		fieldType = mysql.TypeDatetime
+	}
+	resultField = types.NewFieldType(fieldType)
+	resultField.Decimal = mysql.MaxFsp
+	value, err := types.Convert(nodeDate, resultField)
+	if err != nil {
+		e.err = ErrInvalidOperation.Gen("DateArith invalid args, need date but get %T", nodeDate)
+		return false
+	}
+	if types.IsNil(value) {
+		e.err = ErrInvalidOperation.Gen("DateArith invalid args, need date but get %v", value)
+		return false
+	}
+	result, ok := value.(mysql.Time)
+	if !ok {
+		e.err = ErrInvalidOperation.Gen("DateArith need time type, but got %T", value)
+		return false
+	}
+	// parse interval
+	var interval string
+	if strings.ToUpper(v.Unit) == "DAY" {
```

Comment:
@hhkbp2 I mean we can use `v.Form == ast.DateArithDaysForm`  instead of `strings.ToUpper(v.Unit) == "DAY"`.
But your code is better， because I can remove `DateArithDaysForm` later.